forward_warp/forward_warp.py

```diff
@@ -0,0 +1,191 @@
+import tensorflow as tf
+from utils.misc import print_tensor_shape
+
+
+def forward_warp(features, flow):
+    """
+    For an algorithm that gives the same end result, see section 3 in https://arxiv.org/pdf/1711.05890.pdf.
+    Note that the actual implementation here is not n^2, and should be linear in GPU memory.
+    :param features: A Tensor. Features to be warped, of shape [batch_size, H, W, C].
+    :param flow: A Tensor. Un-normalized flow in image pixel units, of shape [batch_size, H, W, 2].
+    """
+
+    # Flip (x, y) to (y, x) for flow, to avoid further confusion.
+    flow = tf.reverse(flow, axis=[-1])
+
+    # Get target indices along with corresponding values to add.
+    indices, splat_values = get_translated_pixels(features, flow)
+
+    # Mask out out-of-bounds splat values
+    height, width, channels = tf.shape(features)[1], tf.shape(features)[2], tf.shape(features)[3]
+    greater_than_zero = tf.greater_equal(indices, 0)
+    greater_than_zero = tf.cast(tf.reduce_all(greater_than_zero, axis=-1), tf.float32)
+    less_than_height = tf.cast(tf.less(indices[..., 0], height), tf.float32)
+    less_than_width = tf.cast(tf.less(indices[..., 1], width), tf.float32)
+    within_bounds = greater_than_zero * less_than_height * less_than_width
+    splat_values *= tf.expand_dims(within_bounds, axis=-1)
+
+    # Clip indices and flatten.
+    y_indices = tf.clip_by_value(indices[..., 0], 0, height-1)
+    x_indices = tf.clip_by_value(indices[..., 1], 0, width-1)
+    indices = tf.stack([y_indices, x_indices], axis=-1)
+
+    # Scatter into output.
+    all_warped = []
+
+    def _scatter(elms):
+        splat_values, indices, features = elms
+        warped = tf.scatter_nd(indices, splat_values, tf.shape(features))
+        all_warped.append(warped)
+        summed = tf.add_n(all_warped)
+
+        # map_fn requires us to return the same number of things as arguments (nested structure must match).
+        # See: https://stackoverflow.com/questions/47984876/tensorflow-tf-map-fn-parameters
+        return summed, summed, summed
+
+    warped, _, _ = tf.map_fn(_scatter, (splat_values, indices, features), back_prop=True)
```

Comment:
What is this map_fn for? I am confused by what it's doing -- why is there an add_n on what looks to be a shared all_warp array? Please add comments.

Reply:
I was cleaning up the code in this section and accidentally left behind some old code. I will fix this and add comments.